### PR TITLE
GUI: Redscreen

### PR DESCRIPTION
**Pull request by michalrudolf**
BFW-2856

Reimplementing redscreen with correct use of out GUIAPI.

### CODE REVIEW

File: src/gui/screen_qr_error.cpp [hunk 1]
```diff
@@ -4,36 +4,125 @@
...
-    info.SetBackColor(COLOR_RED_ALERT);
-    info.SetAlignment(Align_t::Center());
-    static const char hlp[] = "help.prusa3d.com";
-    info.SetText(string_view_utf8::MakeCPUFLASH((const uint8_t *)hlp));
+    : AddSuperWindow<screen_reset_error_data_t>()
+    , err_title(this, title_rect, is_multiline::no)
+    , err_description(this, descr_rect, is_multiline::yes)
+    , hand_icon(this, hand_rect, IDR_PNG_hand_qr)
+    , qr(this, QR_rect, 1) // error code is passed in the constructor
+    , help_txt(this, Rect16(0, 0, 0, 0), is_multiline::no)
```
Comment (RadekVana) [inline src/gui/screen_qr_error.cpp:33]: erase help_txt

File: src/gui/screen_qr_error.cpp [hunk 2]
```diff
@@ -4,36 +4,125 @@
...
+static const constexpr Rect16 link_rect = Rect16(0, 270, display::GetW(), 13);
+static const constexpr Rect16 qr_code_rect = Rect16(100, 295, 64, 13);
+static const constexpr Rect16 fw_version_rect = Rect16(6, 295, 80, 13);
+static const constexpr Rect16 signature_rect = Rect16(160, 295, 40, 13);
+static const constexpr Rect16 appendix_rect = Rect16(195, 295, 40, 13);
+
+static constexpr const char *const header_label = N_("ERROR");
+static constexpr const char *const help_text = N_("More detail at");
 
 screen_qr_error_data_t::screen_qr_error_data_t()
```
Comment (RadekVana) [inline src/gui/screen_qr_error.cpp:27]: use better name, something like ScreenErrorQR

File: src/gui/screen_qr_error.cpp [hunk 3]
```diff
@@ -4,36 +4,125 @@
...
+            printerCode(p_code);
+            qr_code_txt.SetText(string_view_utf8::MakeRAM((const uint8_t *)p_code));
+        } else {
+            qr_code_txt.Hide();
+        }
+    }
+
+    /// draw footer information
+    /// fw version, hash, [fw signed], [appendix]
+    static char fw_version[14]; // intentionally limited to the number of practically printable characters without overwriting the nearby hash text
```
Comment (RadekVana) [inline src/gui/screen_qr_error.cpp:94]: don't use 14 directly
is not size of that string defined somewhere?
If not, you should define it and use it everywhere fw version is accessed. Even if you had to define new header with just a single constant

**Review by RadekVana — CHANGES_REQUESTED** [review]

File: src/gui/screen_qr_error.cpp (hunk 3 above)
Comment (michalrudolf) [reply src/gui/screen_qr_error.cpp]: `6d270b9`

File: src/gui/screen_qr_error.cpp [hunk 4]
```diff
@@ -4,36 +4,125 @@
...
+
+    uint16_t error_code_short = dump_in_xflash_get_code();
+    uint16_t error_code = ERR_PRINTER_CODE * 1000 + error_code_short;
+    uint32_t i = 0;
+    uint32_t count = sizeof(error_list) / sizeof(err_t);
+
+    while (i < count && error_code_short != error_list[i].err_num) {
+        ++i;
+    }
+    if (i == count) {
```
[the remark this answers is not in the document]
Comment (michalrudolf) [reply src/gui/screen_qr_error.cpp:66]: `6d270b9`

File: src/gui/screen_qr_error.cpp (hunk 2 above)
Comment (michalrudolf) [reply src/gui/screen_qr_error.cpp]: `6d270b9`

**Review by RadekVana — APPROVED** [review]

**Pull request merged**